activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -213,7 +213,7 @@ def size
         if !find_target?
           loaded! unless loaded?
           target.size
-        elsif @association_ids
+        elsif @association_ids && target.empty?
           @association_ids.size
```

Comment:
The intent that removing `target.empty?` is that `target.empty?` should be redundant if staled `@association_ids` is properly cleared. But I missed `CollectionAssociation#build` at #29631.

Can you move `@association_ids = nil` after the `target << record` line in `replace_on_target`?

```diff
diff --git a/activerecord/lib/active_record/associations/collection_association.rb b/activerecord/lib/active_record/associations/collection_association.rb
index 68f53d5c1c..77cec85a64 100644
--- a/activerecord/lib/active_record/associations/collection_association.rb
+++ b/activerecord/lib/active_record/associations/collection_association.rb
@@ -347,7 +347,6 @@ def _create_record(attributes, raise = false, &block)
               add_to_target(record) do
                 result = insert_record(record, true, raise) {
                   @_was_loaded = loaded?
-                  @association_ids = nil
                 }
               end
               raise ActiveRecord::Rollback unless result
@@ -424,7 +423,6 @@ def concat_records(records, raise = false)
               unless owner.new_record?
                 result &&= insert_record(record, true, raise) {
                   @_was_loaded = loaded?
-                  @association_ids = nil
                 }
               end
             end
@@ -448,6 +446,7 @@ def replace_on_target(record, index, skip_callbacks)
             target[index] = record
           elsif @_was_loaded || !loaded?
             target << record
+            @association_ids = nil
           end
 
           callback(:after_add, record) unless skip_callbacks
```

Reply:
I think that case doesn't need the `@association_ids = nil`, since the `index` is used for replacing with the same record.

https://github.com/rails/rails/blob/19dc2edaab62864ea96205b78153eed30f752144/activerecord/lib/active_record/associations/collection_association.rb#L283-L288

https://github.com/rails/rails/blob/19dc2edaab62864ea96205b78153eed30f752144/activerecord/lib/active_record/associations/collection_association.rb#L411-L417